lib/Sema/Constraints.cpp

```diff
@@ -0,0 +1,570 @@
+    return new (mem)
+        Constraint(kind, restriction, first, second, locator, typeVars);
+}
+
+Constraint *Constraint::createBindOverload(
+    llvm::BumpPtrAllocator &allocator, glu::types::Ty type,
+    glu::ast::FunctionDecl *choice, glu::ast::ASTNode *locator
+)
+{
+    // Collect type variables.
+    llvm::SmallPtrSet<glu::types::TypeVariableTy *, 4> typeVars;
+
+    getTypeVariable(type, typeVars);
+    getTypeVariable(choice->getType(), typeVars);
+
+    // Create the constraint.
+    auto size = totalSizeToAlloc<glu::types::TypeVariableTy *>(typeVars.size());
+    void *mem = allocator.Allocate(size, alignof(Constraint));
+    return new (mem) Constraint(type, choice, locator, typeVars);
+}
+
+Constraint *Constraint::createDisjunction(
+    llvm::BumpPtrAllocator &allocator, llvm::ArrayRef<Constraint *> constraints,
+    glu::ast::ASTNode *locator, bool rememberChoice
+)
+{
+    // Unwrap any disjunctions inside the disjunction constraint; we only allow
+    // disjunctions at the top level.
+    llvm::SmallPtrSet<glu::types::TypeVariableTy *, 4> typeVars;
+    bool unwrappedAny = false;
+    llvm::SmallVector<Constraint *, 1> unwrapped;
+    unsigned index = 0;
+    for (auto constraint : constraints) {
+        // Gather type variables from this constraint.
+        gatherReferencedTypeVars(constraint, typeVars);
+
+        // If we have a nested disjunction, unwrap it.
+        if (constraint->getKind() == ConstraintKind::Disjunction) {
+            // If we haven't unwrapped anything before, copy all of the
+            // constraints we skipped.
+            if (!unwrappedAny) {
+                unwrapped.append(
+                    constraints.begin(), constraints.begin() + index
+                );
+                unwrappedAny = true;
+            }
+
+            // Add all of the constraints in the disjunction.
+            unwrapped.append(
+                constraint->getNestedConstraints().begin(),
+                constraint->getNestedConstraints().end()
+            );
+        } else if (unwrappedAny) {
+            // Since we unwrapped constraints before, add this constraint.
+            unwrapped.push_back(constraint);
+        }
+        ++index;
+    }
+
+    // If we unwrapped anything, our list of constraints is the unwrapped list.
+    if (unwrappedAny)
+        constraints = unwrapped;
+
+    assert(!constraints.empty() && "Empty disjunction constraint");
+
+    // If there is a single constraint, this isn't a disjunction at all.
+    if (constraints.size() == 1) {
+        assert(!rememberChoice && "simplified an important disjunction?");
+        return constraints.front();
+    }
+
+    // Create the disjunction constraint.
+    auto size = totalSizeToAlloc<glu::types::TypeVariableTy *>(typeVars.size());
+    void *mem = allocator.Allocate(size, alignof(Constraint));
+    auto disjunction = new (mem)
+        Constraint(ConstraintKind::Disjunction, constraints, locator, typeVars);
```

Comment:
you are storing an ArrayRef which is a reference to a stack-allocated array which will be deallocated at the end of the function. this will ofc crash. allocate the `constraints` array within the `allocator` so it gets deallocated at the same time as the rest.